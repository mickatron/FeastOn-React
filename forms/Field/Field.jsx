import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import uniqueId from 'lodash/uniqueId';

import uiWrapper from '../../HOC/uiWrapper';
import Label from '../Label';
import Message from '../../elements/Message';

import classNameType from 'js/shared/proptypes/className';

class Field extends React.PureComponent {

  static propTypes = {
    disabled: PropTypes.bool,
    readOnly: PropTypes.bool,
    description: PropTypes.string,
    label: PropTypes.shape(label.propTypes),
    required: PropTypes.bool,
    valid: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.node,
    ]),
    invalid: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.node,
    ]),
    info: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.node,
    ]),
    customRender:PropTypes.func,
  }

  static defaultProps = {
    showDescription: true,
  }

  componentWillMount() {
    // generate a unique ID for the component
    // TODO: test and update for isomorphic applications
    this.uid = uniqueId();
  }

  uid = undefined;

  render() {
    const {
      className,
      disabled,
      readOnly,
      description,
      info,
      label,
      // input: {component, ...otherInputProps},
      required,
      invalid,
      customRender,
    } = this.props;
    const uid = this; 
    // TODO: refactor elsewhere
    if (disabled) label.input.disabled = true;
    if (required) label.input.required = true;
    if (readOnly) label.input.readOnly = true;

    return (customRender && customRender(this.props)) || (
      <React.Fragment>
        <Label {...label} for={uid} />
        {description? <span className="c-Field__description">{description}</span> : null}
        {info && ( // TODO: test if it's a react node
          <Message modifier="info">
            {info}
          </Message>
        )}
        {invalid && ( // TODO: test if it's a react node
          <Message modifier="falsy">
            {invalid}
          </Message>
        )}
      </React.Fragment>
    );
  }
}

export default uiWrapper({
  name: 'Field',
  Component: Field,
  element: 'Field',
  modifiers: [
    'disabled',
    'readOnly',
    'required',
    'valid',
    'invalid',
    'warning',
    'info',
  ], 
});
