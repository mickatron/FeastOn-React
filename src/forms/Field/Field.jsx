import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import classNameType from 'js/shared/proptypes/className';
import FieldDecorator from '../FieldDecorator/FieldDecorator';

export default class Field extends React.PureComponent {
    
  static propTypes = {
    className: classNameType,
    children: PropTypes.node.isRequired,
    disabled: PropTypes.bool,
    description: PropTypes.string,
    help: PropTypes.string,
    label: PropTypes.string.isRequired,
    isRequired: PropTypes.bool,
    // showHelp: PropTypes.bool,
    showDescription: PropTypes.bool,
    style: PropTypes.object,
    isValid: PropTypes.shape({
      valid: PropTypes.bool,
      warning: PropTypes.bool,
      info: PropTypes.bool,
    }),
    isValidMessage: PropTypes.string,
  }

  static defaultProps = {
    showDescription: true, 
    isValid: {
      valid: undefined,
    },
  }

  render() {
    const { 
      className, 
      disabled, 
      children,
      description,
      help,
      label,
      showDescription,
      style,
      isRequired,
      isValid,
      isValidMessage,
    } = this.props;
    
    return ( 
      <FieldDecorator
        className={classnames('c-Field', className)}
        disabled={disabled}
        isValid={isValid}
        required={isRequired}
        style={style}
      >
        <label
          className="c-Field__label"
        >{label}</label>
        {help && isValid.info ? <span className="c-Field__help-text">{help}</span> : null}
        {children}
        {description && showDescription ? <span className="c-Field__description">{description}</span> : null}
        {isValid.valid === false ? <p className="c-Field__errorMessage">{isValidMessage}</p> : null }
      </ FieldDecorator>
    );
  }

}
