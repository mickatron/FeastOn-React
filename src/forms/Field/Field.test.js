import { shallow } from 'enzyme';
import React from 'react';

import Field from './Field';
// import data from '../mock/data'

describe('Field', () => { 
  let component;
  describe('renders', () => { 
    it('with label and custom classname', ()=>{
      component = shallow(
        <Field
          label="form label"
          className="test"
        >
          <input />
        </ Field>
      );
      expect(component).toMatchSnapshot();
      expect(component.find('.c-Field.test').exists()).toBeTruthy();
    });

    it('with label and custom classname', ()=>{
      component = shallow(
        <Field
          label="form label"
          className="test"
          helpText="some test help text"
          valid={false}
          validationMessage="some test validation message"
        >
          <input />
        </ Field>
      );
      expect(component).toMatchSnapshot();
      expect(component.find('.c-Field.test').exists()).toBeTruthy();
    });

  });
});