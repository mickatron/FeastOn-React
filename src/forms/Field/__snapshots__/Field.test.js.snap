// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field renders with label and custom classname 1`] = `
<FieldDecorator
  className="c-Field test"
  isValid={
    Object {
      "valid": undefined,
    }
  }
>
  <label>
    form label
  </label>
  <input />
</FieldDecorator>
`;

exports[`Field renders with label and custom classname 2`] = `
<FieldDecorator
  className="c-Field test"
  isValid={
    Object {
      "valid": undefined,
    }
  }
>
  <label>
    form label
  </label>
  <input />
</FieldDecorator>
`;
